Bluetooth: Host: Re-order Dynamic LE L2CAP Credit-based connection accept procedure

donatieng opened the pull request:
Currently, when an incoming dynamic LE L2CAP connection is requested by a peer, the `connected()` callback provided by the user is raised *before* a `L2CAP_LE_CREDIT_BASED_CONNECTION_RSP` is sent back to the peer.

In some cases the user will start sending data in the `connected()` callback which would be received too early by the peer. 

This commit/PR fixes this behavior by making sure the `connected()` callback is raised only after the connection response has been sent to the peer.

Note: This is my first Zephyr PR so I welcome any guidance on how best to shape it/deliver this fix!

donatieng committed:
Bluetooth: Host: Re-order LE L2CAP connection accept procedure

Currently, when an incoming dynamic LE L2CAP connection is requested by
a peer, the connected() callback provided by the user is raised
*before* a L2CAP_LE_CREDIT_BASED_CONNECTION_RSP is sent back to the
peer.
In some cases the user will start sending data in the connected()
callback which would be received too early by the peer.
This commit fixes this behavior by making sure the connected() callback
is raised only after the connection response has been sent to the peer.

Signed-off-by: Donatien Garnier <donatien.garnier@blecon.net>

jori-nordic commented:
Could you please add a test for that? 
- device A busy-waits during the connected callback, holding up the `RSP`
- device B starts sending
- device A stops busy waiting, checks that all the data has arrived (no packets missed)

hermabe commented:
There was some related discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/55408, where it was also suggested to move the connected callback later.

theob-pro commented:
Hey @donatieng, if you need help implementing the test feel free to contact me on Discord.

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1310,6 +1312,13 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 		ecred_cb->ecred_conn_req(conn, result, psm);
 	}
+
+	for (i = 0; i < req_cid_count; i++) {
+		/* Raise connected callback for established channels */
+		if ((dcid[i] != 0x00) && (chan[i]->ops->connected != NULL)) {
+			chan[i]->ops->connected(chan[i]);
+		}
+	}
```
I think the connected callback should not fire until/unless a conn_rsp PDU with BT_L2CAP_LE_SUCCESS has been queued on the connection TX queue.

```suggestion
	if (rsp_queued && result == BT_L2CAP_LE_SUCCESS) {
		for (i = 0; i < req_cid_count; i++) {
			/* Raise connected callback for established channels */
			if ((dcid[i] != 0x00) && (chan[i]->ops->connected != NULL)) {
				chan[i]->ops->connected(chan[i]);
			}
		}
	}
```

`rsp_queued` is not defined yet. I suggest manually inlining `l2cap_send` and testing the return code of `bt_l2cap_send`. A similar inlining should be done in `le_conn_req`.

donatieng replied:
Makes sense @alwa-nordic, thanks for spotting this - I'll amend the PR

alwa-nordic reviewed (changes requested):
Thank you for your PR! It's a good improvement :). I have a comment on handling a failure condition that I think should be addressed.

donatieng commented:
@hermabe thanks for pointing out #55408 - I hadn't realised a similar PR had been raised, so it was useful reading back through it (and it seems that this PR is in line with the discussion's outcomes).

@jori-nordic In terms of the test not sure if what you're suggesting makes sense to me as B won't be able to send data before having receive RSP. I assume A is acting as the L2CAP server, B as the L2CAP client.
Here's the flow I assume would fail before this PR, and succeed after - so would propose to write the following test:
* B makes a connection request
* A accepts
* A sends data in the connected() callback
* B waits for connected() callback and the data to be received

Thanks @theob-pro for offering help with the test, I might take you up on that!

jori-nordic commented:
@donatieng that's fine, I just want the behavior tested so we don't introduce a regression later :+1:

donatieng commented:
No problem - thanks for the quick feedback all, will try and make progress in the coming days

donatieng committed:
Bluetooth: Host: Add sending L2CAP data from connection callback test

Add BabbleSim test to make sure data can be sent immediately upon
connection from the L2CAP channel connection callback

Signed-off-by: Donatien Garnier <donatien.garnier@blecon.net>

donatieng commented:
I've now added a BabbelSim test which should fail prior to the introduction of the second commit, and addressed suggestions above

theob-pro commented on tests/bsim/bluetooth/host/l2cap/send_on_connect/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(bsim_test_l2cap_send_on_connect)
+
+FILE(GLOB app_sources src/*.c)
```
```suggestion
FILE(GLOB app_sources src/*.c)
target_sources(app PRIVATE ${app_sources})

target_sources(app PRIVATE
    main.c
    common.c
    main_l2cap_send_on_connect.c
)
```
We try to avoid usage of globbing.

donatieng replied:
No problem, fixed :)

jori-nordic reviewed (changes requested):
thanks for the test and patch, I only have one nitpick: the new label is not necessary for the non-ecred path.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1193,7 +1190,15 @@ static void le_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 
 rsp:
-	l2cap_send(conn, BT_L2CAP_CID_LE_SIG, buf);
+	if (bt_l2cap_send(conn, BT_L2CAP_CID_LE_SIG, buf)) {
```
There seems to be inconsistency with using `result` vs `rsp->result` in this function. In some cases the value of `result` gets ignored and some other value gets encoded into `rsp->result`. Since you're updating the code flow, I'd suggest to fix this so that there's a single place that encodes `rsp->result`, right after the `rsp:` label:
```c
	rsp->result = sys_cpu_to_le16(result);
```

So that means that all the places with `goto rsp;` should set `result` instead of `rsp->result`.

donatieng commented:
@jhedberg I was getting worried this refactor was heading towards a bit of scope creep, however having looked back at this function and the rest of the file, this issue seems quite self contained and hence it's a good opportunity for a quick refactor. Just pushed the changes 🙂

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1144,6 +1140,7 @@ static void le_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	mtu = sys_le16_to_cpu(req->mtu);
 	mps = sys_le16_to_cpu(req->mps);
 	credits = sys_le16_to_cpu(req->credits);
+	result = BT_L2CAP_LE_ERR_PSM_NOT_SUPP;
```
This seems like an unnecessary assignment, unless I overlooked something: there is no branch that doesn’t set the value which also ends up using the value.

donatieng replied:
Good catch! Fixed 🙂

jhedberg reviewed (changes requested):
Looks pretty good now, thanks! Just one minor thing I noticed (see my inline comment)

donatieng commented:
Thanks everyone for your help getting this PR in!
